shedlock-provider-hazelcast/src/main/java/net/javacrumbs/shedlock/provider/hazelcast/HazelcastLockProvider.java

```diff
@@ -0,0 +1,191 @@
+ */
+public class HazelcastLockProvider implements LockProvider {
+
+    private static final Logger log = LoggerFactory.getLogger(HazelcastLockProvider.class);
+
+    protected static final String LOCK_STORE_KEY_DEFAULT = "shedlock_storage";
+
+    /**
+     * Key used for get the lock container (an {@link IMap}) inside {@link #hazelcastInstance}.
+     * By default : {@link #LOCK_STORE_KEY_DEFAULT}
+     */
+    protected final String lockStoreKey;
+
+    /**
+     * Instance of the Hazelcast engine used by the application.
+     */
+    protected HazelcastInstance hazelcastInstance;
+
+    /**
+     * Instanciate the provider.
+     *
+     * @param hazelcastInstance The Hazelcast engine used by the application.
+     */
+    public HazelcastLockProvider(final HazelcastInstance hazelcastInstance) {
+        this(hazelcastInstance, LOCK_STORE_KEY_DEFAULT);
+    }
+
+    /**
+     * Instanciate the provider.
+     *
+     * @param hazelcastInstance The Hazelcast engine used by the application
+     * @param lockStoreKey      The key where the locks store is associate {@link #hazelcastInstance} (by default {@link #LOCK_STORE_KEY_DEFAULT}).
+     */
+    public HazelcastLockProvider(final HazelcastInstance hazelcastInstance, final String lockStoreKey) {
+        this.hazelcastInstance = hazelcastInstance;
+        this.lockStoreKey = lockStoreKey;
+    }
+
+    @Override
+    public Optional<SimpleLock> lock(final LockConfiguration lockConfiguration) {
+        log.trace("lock - Attempt : {}", lockConfiguration);
+        final Instant now = Instant.now();
+        final String lockName = lockConfiguration.getName();
+        final IMap<String, HazelcastLock> store = getStore();
+        try {
+            // lock the map key entry
+            store.lock(lockName);
+            // just one thread at a time, in the cluster, can run this code
+            // each thread waits until the lock to be unlock
+            if (tryLock(lockConfiguration, now)) {
+                return Optional.of(() -> unlock(lockName));
+            }
+        } finally {
+            // released the map lock for the others threads
+            store.unlock(lockName);
+        }
+        return Optional.empty();
+    }
+
+    private boolean tryLock(final LockConfiguration lockConfiguration, final Instant now) {
+        final String lockName = lockConfiguration.getName();
+        final HazelcastLock lock = getLock(lockName);
+        if (isUnlocked(lock)) {
+            log.debug("lock - lock obtained, it wasn't locked : conf={}", lockConfiguration);
+            addNewLock(lockConfiguration);
+            return true;
+        } else if (isExpired(lock, now)) {
+            log.debug("lock - lock obtained, it was locked but expired : oldLock={};  conf={}", lock, lockConfiguration);
+            removeLock(lockName);
```

Comment:
(Minor) replaceLock method would have been nice to prevent this small code duplication.